templates/extras/simp.md

```diff
@@ -63,221 +74,251 @@ lemmas". Here are some examples of `simp` lemmas in mathlib:
+innermost terms, working outward: it first simplifies the arguments of
+a function before simplifying the function. Also, `simp` contains a some
+amount of cleverness to be able to avoid considering *all* `simp`
+lemmas every time (there are over ten thousand of them currently in `mathlib`!).
+
+The simplifier works in one direction only: if `A = B` is a `simp`
+lemma, then `simp` replaces `A`s with `B`s, but it doesn't replace
+`B`s with `A`s. Hence a `simp` lemma should have the property that its
+right-hand side should be simpler than its left-hand side. In
 particular, `=` and `↔` should not be viewed as symmetric operators in
-this situation. The following lemma
+this situation. The following would be a terrible `simp` lemma (if it
+were even allowed):
 
 ```lean
 @[simp] lemma mul_right_inv_bad (a : G) : 1 = a * a⁻¹ := ...
 ```
 
-would be a terrible `simp` lemma; the simplifier would then attempt to
-replace `1` with `a * a⁻¹`, which is usually not a sensible direction
-of travel. Furthermore, if both `mul_right_inv` and
-`mul_right_inv_bad` are tagged with `simp` then this would cause the
-simplifier to loop.
+Replacing `1` with `a * a⁻¹` is not a sensible default direction to
+travel. Even worse would be a lemma that causes expressions to grow
+without bound, causing `simp` to loop forever:
 
-When making a new definition, it is very common to have `@[simp]`
-lemmas which train the simplifier how to use the definition in a
-sensible manner. An example of this is in mathlib's
+```lean
+@[simp] lemma even_worse_lemma: (1 : G) = 1 * 1⁻¹ := ...
+```
+
+When making a new definition, it is very common to also introduce
+`simp` lemmas to put expressions involving the definition into a
+sensible form. An example of this is in mathlib's
 [data.complex.basic](https://github.com/leanprover-community/mathlib/blob/master/src/data/complex/basic.lean),
-which has almost 100 `simp` lemmas in. Theorems such as `@[simp] lemma
-add_re (z w : ℂ) : (z + w).re = z.re + w.re := rfl` are proved in that
-file, even though they are true by definition, because they are
-important `simp` lemmas. The very slick proof in that file that the
-complex numbers are a commutative ring uses `simp` at a key moment to
-reduce goals such as `(z + w).re = (w + z).re` into `z.re + w.re =
-w.re + z.re`, which is a statement about real numbers which the `ring`
-tactic can solve.
+which has almost 100 `simp` lemmas. Even though they are true by definition, theorems such as
+```lean
+@[simp] lemma
+add_re (z w : ℂ) : (z + w).re = z.re + w.re := rfl
+```
+are introduced because they give `simp` the ability to reduce expressions and then make use of pre-existing facts.
+This one, for example, converts complex addition into real addition.
+If you give `simp` permission to use commutativity of real addition, then it is able to
+automatically prove `(z + w).re = (w + z).re` through `z.re + w.re =
+w.re + z.re`, which is half of the proof that complex addition is commutative.
 
```

Comment:
Maybe add here something like:

```suggestion
Anytime you create a new definition in Lean, you should also write some appropriate `simp` lemmas.
This will make it easier for others to use your definition without having to unfold it.

If you are defining something which is a structure, putting the `@[simps]` attribute on the definition
will automatically create `simp` lemmas for the projections.
```

Reply:
Thanks for the suggestion. I tried explaining how it's useful having simp lemmas that do partial evaluation, and how `@[simps]` can help.